enhancements/cluster-logging/container-log-throttling.md

```diff
@@ -0,0 +1,206 @@
+
+## Open Questions
+
+1. How can we help cluster administrators identify containers that require log throttling?
+1. How can groupings of containers can be expressed in OpenShift for the purposes of log throttling?
+
+### Goals
+
+* Ship an OpenShift release with an implementation of the container logging policies and rate limiting functionality described in this enhancement while meeting quality goals.
+
+### Non-Goals
+
+* It is not a goal of this enhancement to introduce control of logs from non-containerized OCP components.
+* It it not a goal to introduce log throttling into all container runtimes and container and VM technologies that can be run on Kubernetes. The scope is limited to the ones necessary to enable shipping of the next version OpenShift: Kubernetes API, kubelet, cri-o, podman and conmon on Linux.
+
+## Proposal
+
+### User Stories
+
+#### As a cluster administrator, I want to disable logs from a particular container or a group of containers.
+
+#### As a cluster administrator, I want to cap the rate at which a container or a group of containers can produce logs. I want to keep all the log entries and I don't mind slowing down the containers for that.
+
+#### As a cluster administrator, I want to cap the rate at which a container or a group of containers can produce logs. I don't want to slow down the containers and I don't mind losing some of the log entries.
+
+### Risks and Mitigations
+
+1. This proposal calls for modifications to core Kubernetes APIs and components and as such runs the risk of requiring substantial investment into building consensus around it in the Kubernetes community.
+1. It may take a substantial amount of time for the proposed changes to make it into released versions of the corresponding components.
+
+## Design Details
+
+The general idea of the design is to build the container log throttling mechanism into [conmon](https://github.com/containers/conmon), the container runtime monitor, and expose the controls successively through the CRI interface and the Container core/v1 API. A POC for the conmon part of the implementation has been built, please refer to the see-also field in the header of this document.
+
+It is envisioned that the following changes will need to be made to the Kubernetes API
+
+### Container core/v1 API:
+
+```go
+// k8s.io/api/core/v1/types.go
+
+// A single application container that you want to run within a pod.
+type Container struct {
+...
+	// Policies the container runtime applies to this container's
+	// log streams. LogStreams[0] controls stdout, LogStreams[1]
+	// controls stderr.
+	// Cannot be updated.
+	// +optional
+    // +patchMergeKey=logStream
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=logStream
+	LogPolicies []LogStreamPolicy `json:"logPolicies,omitempty" patchStrategy:"merge" patchMergeKey:"logStream" protobuf:"bytes,23,rep,name=logPolicies"`
```

Comment:
What does it mean to apply multiple polices to a single container?  Wouldn't there only be one?

Reply:
There's one per log stream, stdout can be conrolled separately from stderr.